Add benchmark:{run, diff} mage targets

alexsapran opened the pull request:
This PR adds 2 new mage targets for the project that execute the go bench benchmarks and benchstat for comparing the results.

Example use of the above is:
```
$ mage benchmark:run benchmark-before
# ..... make changes ......
$ mage benchmark:run benchmark-after
$ mage benchmark:diff benchmark-before benchmark-after
```

Once we are happy with the workflow we can move this part into the `elastic-agent-libs/dev-tools/mage` and reference from there.

The intention of these new mage targets is to be used when adding code to this repo with the above workflow and to integrate them into a GH action and have it report the numbers back to the PR, an example workflow that I was playing around with can be found in https://github.com/alexsapran/elastic-agent-shipper-client/pull/1#issuecomment-1450737792

alexsapran committed:
Add benchmark:run and benchmark:diff mage targets

This PR adds 2 new mage targets for the project that execute the go
bench benchmarks and benchstat for comparing the results.

Example use of the above is:
```
$ mage benchmark:run benchmark-before
$ mage benchmark:run benchmark-after
$ mage benchmark:diff benchmark-before benchmark-after
```

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-client-mbp%2FPR-22/detail/PR-22/10//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-client-mbp%2FPR-22/detail/PR-22/10//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-client-mbp%2FPR-22/detail/PR-22/10//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent-shipper-client-mbp%2FPR-22/detail/PR-22/10//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent-shipper-client_22.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-03-15T19:59:54.134Z&rangeTo=2023-03-15T20:19:54.134Z&transactionName=BUILD+elastic-agent-shipper-client-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=d8dbe72e2ecb8e2d1e2c845207f8cdb1&transactionId=50e50093cf72b713)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-03-15T20:09:54.134+0000


* Duration: 3 min 55 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 21  |
| Skipped      | 0 |
| Total        | 21   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

faec commented:
Sorry, I'm lacking some context -- this repository is for boilerplate API definitions used to make RPC calls to the shipper, it doesn't run anything by itself. What's being benchmarked?

alexsapran commented:
Hi @faec, you were quicker to comment than me writing all the context. 

We introduced micro benchmarks in https://github.com/elastic/elastic-agent-shipper-client/pull/17 and intended to have them as part of the workflow of future changes. 
I will be pushing another commit as well to add the GH action that would use the new mage targets. I am asking you to hold reviewing until I am finished. You were automatically got assigned while opening the PR :D

alexsapran committed:
Fix the go.mod

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

alexsapran commented:
I think I am done with the changes in this PR, I would have to do a follow-up that only adds the gh action that is going to be using those targets.

cmacknz commented on magefile.go:
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
+	if outputFile != "" {
+		fileOutput, err := os.Create(createDir(outputFile))
+		if err != nil {
+			return errors.Wrap(err, "failed to create go test output file")
```
You want to use the standard library's `errors` package with the `%w` wrapping verb instead:

```suggestion
			return fmt.Errorf("failed to create go test output file: %w", err)
```

We have a lint warning in the other repositories that suggest it but we must be missing it here. The pkg/errors version is not actively maintained anymore and isn't exactly compatible with the standard library version.

alexsapran replied:
Thanks for the info, I was not aware of this. I used a pre-existing code from `beats`, I am guessing that lint has a bug?

cmacknz replied:
Not a bug in the linter, the Beats code was just written before the standard library included the features from pkg/errors and rewriting all of it will take a long time.

cmacknz commented on magefile.go:
```diff
@@ -8,15 +8,22 @@
 	"github.com/elastic/elastic-agent-libs/dev-tools/mage/gotool"
 	"github.com/magefile/mage/mg"
 	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
```
Replace all uses of this with the stdlib's `errors` package

cmacknz commented:
We implement common mage targets in https://github.com/elastic/elastic-agent-libs/tree/main/dev-tools/mage if you end up moving these to another repository afterwards and don't want to just copy them.

fearful-symmetry commented on magefile.go:
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
+	err := goTestBench.Run()
+
+	var goTestErr *exec.ExitError
+	if err != nil {
```
I'm not sure why the error checks are in two separate if statements, it's a tad hard to follow at first. It might be a little more concise if we just do something like `if errors.Is() {...} else if errors.Is() {...} else err != nil {...}` or something like that.

alexsapran replied:
I had not put too much thought into this part of the build as my main focus was the github action.
It is a copy/paste from `beats` https://github.com/elastic/beats/blob/c3f5d7eed9405f71fea618ffa53e895c3c908aff/dev-tools/mage/gotest.go#L271-L278

I will try your suggestion 👍🏼

fearful-symmetry commented on magefile.go:
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
+	return nil
+}
+
+func (Benchmark) Run(ctx context.Context, outputFile string) error {
```
I wonder if there's a way to make this a tad more user friendly, since if you omit the output file, you get a rather unhelpful `not enough arguments for target "Benchmark:Run", expected 1, got 0` error message from mage. Just adding a godoc that says something like `// Run all benchmarks. Takes the name of an output file as an argument` might help. Ditto with `benchmark:diff`

alexsapran replied:
please see [66c329c](https://github.com/elastic/elastic-agent-shipper-client/pull/22/commits/66c329c2fd2b0f34ecfd6252edab1906995ddb93) if makes this part any better.

alexsapran committed:
Remove pkg/error and use errors

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

alexsapran committed:
Rework error handing

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

alexsapran committed:
Move parameters to env vars

Now instead of running the mage target with the parameters you would
pass the parameters as environment variables before the target.

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

alexsapran committed:
Revert "Fix the go.mod"

This reverts commit 0b06b069f1cdf89bbc9152a5938715c2570316ec.

fearful-symmetry commented on magefile.go:
```diff
@@ -181,3 +196,144 @@ func licenser(mode licenserMode) error {
+	return nil
+}
+
+func (Benchmark) Diff(ctx context.Context) error {
```
Can we add godocs to the new mage targets? The rest of them have docs, and it's nice for usability/consistency: 
```
  benchmark:diff
  benchmark:run
  check                  runs all the checks
  checkLicenseHeaders    checks ASL2 headers in .go files
  generateGo             regenerates the Go files out of .proto files
  installBenchStat
  installLicenser        installs the go-licenser.
  installProtoGo         installs required plugins for protoc
  license                applies the right license header.
  update                 updates all the generated code out of the spec
```

fearful-symmetry commented on magefile.go:
```diff
@@ -181,3 +196,144 @@ func licenser(mode licenserMode) error {
+		args = append(args, baseFile)
+	}
+	if nextFile == "" {
+		return fmt.Errorf("Missing candidate benchmark output file, please run first OUTPUT=foo benchmark:run")
```
Small nit: can you make the `OUTPUT=foo` line a bit more descriptive? Something like `OUTPUT=path_to_other_benchmark_output` or something like that.

alexsapran committed:
Add docs on the mage targets

Signed-off-by: Alexandros, Sapranidis <alexandros@elastic.co>

alexsapran commented:
I am going to close as I have started the work to migrate this into [elastic-agent-libs](https://github.com/elastic/elastic-agent-libs/pull/110)
